gnovm/pkg/gnolang/nodes.go

```diff
@@ -385,33 +390,65 @@ type BasicLitExpr struct {
 	Value string
 }
 
+func (x *BasicLitExpr) addressability() Addressability {
+	return AddressabilityUnsatisfied
+}
+
 type BinaryExpr struct { // (Left Op Right)
 	Attributes
 	Left  Expr // left operand
 	Op    Word // operator
 	Right Expr // right operand
 }
 
+func (x *BinaryExpr) addressability() Addressability {
+	return AddressabilityUnsatisfied
+}
+
 type CallExpr struct { // Func(Args<Varg?...>)
 	Attributes
-	Func    Expr  // function expression
-	Args    Exprs // function arguments, if any.
-	Varg    bool  // if true, final arg is variadic.
-	NumArgs int   // len(Args) or len(Args[0].Results)
+	Func           Expr  // function expression
+	Args           Exprs // function arguments, if any.
+	Varg           bool  // if true, final arg is variadic.
+	NumArgs        int   // len(Args) or len(Args[0].Results)
+	Addressability Addressability
```

Comment:
For the cases like this where the addressability is assigned during preprocess, what do you think about moving Addressability as an attribute?

The reason why I say this is that we currently have a bunch of information which is not actually persisted, but recovered by using `Preprocess`. The `Expr` types here contain in their struct only information which comes directly out of the AST representation; not from preprocessing. I'm not entirely sure of why it's designed this way; but I think for consistency we should maintain the split between "preprocessed information" in Attributes, and AST information in the struct themselves.

If addressability is an attribute, I also wonder if there's a way to remove the specific type and turn this into `Addressable() bool` instead.

Reply:
I don't think there is a good way to make an `Addressable() bool` method due to reasons I outlined in responses to other comments -- there are one or more cases where addressability is not applicable so the preprocessor should handle execution as usual.  That being said, while many other attributes apply only exist some of the time, the Addressability attribute exists for all addressable expressions -- everywhere it appears.  For ease of access, I'd prefer not to make this an attribute.